mentat/include_files.py

```diff
@@ -44,104 +25,281 @@ def is_file_text_encoded(abs_path: Path):
+            validate_file_interval_path(abs_path, check_for_text)
+        case PathType.DIRECTORY:
+            pass
+        case PathType.GLOB:
+            validate_glob_path(abs_path)
+
+    return abs_path
+
+
+def match_path_with_patterns(path: Path, patterns: Set[str]) -> bool:
+    """Check if the given absolute path matches any of the patterns.
+
+    TODO: enforce valid glob patterns? Right now we allow glob-like
+    patterns (the one's that .gitignore uses). This feels weird imo.
+    """
+    if not path.is_absolute():
+        raise PathValidationError(f"Path {path} is not absolute")
+
+    for pattern in patterns:
+        # Prepend '**' to relative patterns that are missing it
+        if not Path(pattern).is_absolute() and not pattern.startswith("**"):
+            if fnmatch.fnmatch(str(path), str(Path("**").joinpath(pattern))):
+                return True
+        if fnmatch.fnmatch(str(path), pattern):
+            return True
+        for part in path.parts:
+            if fnmatch.fnmatch(str(part), pattern):
+                return True
 
-    # config glob excluded files only apply to files added from directories
-    files_from_dirs = [
-        file
-        for file in files_from_dirs
-        if file.path.resolve() not in glob_excluded_files
-    ]
-    files_direct.update(files_from_dirs)
+    return False
 
-    files: defaultdict[Path, list[CodeFeature]] = defaultdict(list)
-    for file in files_direct:
-        if file.path not in excluded_files:
-            files[file.path.resolve()].append(file)
 
-    return dict(files), invalid_paths
+def get_paths_for_directory(
+    path: Path,
+    include_patterns: Iterable[Path | str] = [],
+    exclude_patterns: Iterable[Path | str] = [],
+    recursive: bool = True,
+) -> Set[Path]:
+    """Get all file paths in a directory.
 
+    Args:
+        `path` - An absolute path to a directory on the filesystem
+        `include_patterns` - An iterable of paths and/or glob patterns to include
+        `exclude_patterns` - An iterable of paths and/or glob patterns to exclude
+        `recursive` - A boolean flag to recursive traverse child directories
 
-def build_path_tree(files: list[Path], git_root: Path):
+    Return:
+        A set of absolute file paths
+    """
+    paths: Set[Path] = set()
+
+    if not path.exists():
+        raise PathValidationError(f"Path {path} does not exist")
+    if not path.is_dir():
+        raise PathValidationError(f"Path {path} is not a directory")
+    if not path.is_absolute():
+        raise PathValidationError(f"Path {path} is not absolute")
+
+    all_include_patterns = set(str(p) for p in include_patterns)
+    all_exclude_patterns = set(str(p) for p in exclude_patterns)
+
+    for root, dirs, files in os.walk(path, topdown=True):
+        root = Path(root)
+
+        if get_git_root_for_path(root, raise_error=False):
+            dirs[:] = []
+            git_non_gitignored_paths = get_non_gitignored_files(root)
```

Comment:
get_non_git_ignore should be recursive already; I think putting inside of os.walk is going to be really slow. If we have a git root, we shouldn't need to do os.walk (just get_non_gitignored_files).

Reply:
^ yup